platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -256,6 +499,66 @@ odp_crypto_session_create(const odp_crypto_session_param_t *param,
 		session->cipher.func = null_crypto_routine;
 		rc = 0;
 		break;
+#if ODP_DEPRECATED_API
+	case ODP_CIPHER_ALG_AES128_GCM:
+		/* AES-GCM requires to do both auth and
+		 * cipher at the same time */
+		if (param->auth_alg != ODP_AUTH_ALG_AES128_GCM) {
+			rc = -1;
+		} else if (param->cipher_key.length == 16) {
+			if (armv8_aes_gcm_set_constants(AES_GCM_128,
+							session->p.auth_digest_len,
+							session->p.cipher_key.data,
+							&session->cc) != 0) {
+				ODP_DBG("ARM Crypto: Failure in setting constants\n");
+				rc = -1;
+			}
+			rc = process_aes_gcm_param(session);
+		} else {
+			rc = -1;
+		}
+
+		break;
+#endif
+	case ODP_CIPHER_ALG_AES_GCM:
+		/* AES-GCM requires to do both auth and
+		 * cipher at the same time */
+		if (param->auth_alg != ODP_AUTH_ALG_AES_GCM) {
+			rc = -1;
+		} else if (param->cipher_key.length == 16) {
+			if (armv8_aes_gcm_set_constants(AES_GCM_128,
+							session->p.auth_digest_len,
+							session->p.cipher_key.data,
+							&session->cc) != 0) {
+				ODP_DBG("ARM Crypto: Failure in setting constants\n");
+				rc = -1;
+			}
+			rc = process_aes_gcm_param(session);
+
+		} else if (param->cipher_key.length == 24) {
+			if (armv8_aes_gcm_set_constants(AES_GCM_192,
+							session->p.auth_digest_len,
+							session->p.cipher_key.data,
+							&session->cc) != 0) {
+				ODP_DBG("ARM Crypto: Failure in setting constants\n");
+				rc = -1;
+			}
+			rc = process_aes_gcm_param(session);
+
+		} else if (param->cipher_key.length == 32) {
+			if (armv8_aes_gcm_set_constants(AES_GCM_256,
```

Comment:
There is quite a bit of duplicated code here. It makes the code harder to read than it needs to be. You could simply have the following switch statement for setting the mode and then have the library call etc just once. And I think the deprecated API case can also be included in that.

```
switch (param->cipher_key.length) {
case 16:
	mode = AES_GCM_128;
	break;
case 24:
	mode = AES_GCM_192;
	break;
case 32:
	mode = AES_GCM_256;
	break;
default:
	rc = -1;
	break:
}
```

Reply:
Okay, I will make the change. Thanks!